File: backend/src/test/java/com/ssafy/yorr/game/yacht/YachtBotGameCompletionTest.java
package com.ssafy.yorr.game.yacht;

import com.ssafy.yorr.game.domain.ScoreBoard;
import com.ssafy.yorr.game.domain.ScoreCategory;
import com.ssafy.yorr.game.round.application.RoundStartedEvent;
import com.ssafy.yorr.game.round.application.RoundSynchronizationService;
import com.ssafy.yorr.game.round.application.RoundTimerService;
import com.ssafy.yorr.game.round.application.ScoreRoundSubmissionService;
import com.ssafy.yorr.game.round.infrastructure.InMemoryRoundStateStore;
import com.ssafy.yorr.game.repository.ScoreBoardStore;
import com.ssafy.yorr.game.service.ScoreConfirmationService;
import com.ssafy.yorr.room.dto.ParticipantKind;
import com.ssafy.yorr.room.dto.RoomPhase;
import com.ssafy.yorr.room.dto.RoomPlayerSnapshot;
import com.ssafy.yorr.room.dto.RoomSnapshot;
import com.ssafy.yorr.room.service.RoomService;
import com.ssafy.yorr.ws.RoomBroadcaster;
import org.junit.jupiter.api.Test;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class YachtBotGameCompletionTest {

    @Test
    void twoLocalBotsCompleteAllTwelveRoundsThroughTheSharedActionPath() {
        InMemoryRoundStateStore stateStore = new InMemoryRoundStateStore();
        RoundSynchronizationService rounds = new RoundSynchronizationService(stateStore);
        RoundTimerService timers = mock(RoundTimerService.class);
        RoomBroadcaster broadcaster = mock(RoomBroadcaster.class);
        RoomService rooms = mock(RoomService.class);
        InMemoryScoreBoardStore scoreStore = new InMemoryScoreBoardStore();
        ScoreConfirmationService scores = new ScoreConfirmationService(scoreStore);
        ScoreRoundSubmissionService submissions =
                new ScoreRoundSubmissionService(rounds, scores, rooms);
        YachtTurnActionService actions =
                new YachtTurnActionService(rounds, timers, broadcaster, submissions);
        YachtBotTurnCoordinator coordinator = new YachtBotTurnCoordinator(
                rounds,
                actions,
                new LocalYachtBotStrategy(),
                rooms,
                scores
        );
        when(rooms.getSnapshot("room-a")).thenReturn(botRoom());
        rounds.initialize("room-a", 1, List.of("bot-easy", "bot-hard"));

        int actionsExecuted = 0;
        while (rounds.findByRoomId("room-a").orElseThrow().isFinished() == false
                && actionsExecuted < 200) {
            var state = rounds.findByRoomId("room-a").orElseThrow();
            assertThat(coordinator.playIfCurrent(new RoundStartedEvent("room-a", state)))
                    .isTrue();
            actionsExecuted++;
        }

        var finished = rounds.findByRoomId("room-a").orElseThrow();
        assertThat(finished.isFinished()).isTrue();
        assertThat(finished.roundNumber()).isEqualTo(12);
        assertThat(actionsExecuted).isBetween(96, 144);
        assertThat(scoreStore.findScoreBoard("game-a", "bot-easy").categories())
                .doesNotContainValue(null);
        assertThat(scoreStore.findScoreBoard("game-a", "bot-hard").categories())
                .doesNotContainValue(null);
    }

    private static RoomSnapshot botRoom() {
        return new RoomSnapshot(
                "room-a",
                YachtDiceGameModule.CODE,
                "game-a",
                "bot-easy",
                RoomPhase.PLAYING,
                6,
                List.of(
                        new RoomPlayerSnapshot(
                                "bot-easy",
                                "Easy Bot",
                                0,
                                ParticipantKind.BOT
                        ),
                        new RoomPlayerSnapshot(
                                "bot-hard",
                                "Hard Bot",
                                0,
                                ParticipantKind.BOT
                        )
                )
        );
    }

    private static final class InMemoryScoreBoardStore implements ScoreBoardStore {

        private final Map<String, Map<ScoreCategory, Integer>> scores = new HashMap<>();

        @Override
        public ScoreBoard confirmScore(
                String gameId,
                String playerId,
                int roundNumber,
                ScoreCategory category,
                int score,
                String requestSignature
        ) {
            Map<ScoreCategory, Integer> playerScores =
                    scores.computeIfAbsent(playerId, ignored -> new EnumMap<>(ScoreCategory.class));
            if (playerScores.putIfAbsent(category, score) != null) {
                throw new IllegalStateException("category already used: " + category);
            }
            return scoreboard(playerScores);
        }

        @Override
        public ScoreBoard findScoreBoard(String gameId, String playerId) {
            return scoreboard(scores.getOrDefault(playerId, Map.of()));
        }

        private static ScoreBoard scoreboard(Map<ScoreCategory, Integer> scores) {
            int upperSubtotal = scores.entrySet().stream()
                    .filter(entry -> entry.getKey().isUpperCategory())
                    .mapToInt(Map.Entry::getValue)
                    .sum();
            int upperBonus = upperSubtotal >= 63 ? 35 : 0;
            int total = scores.values().stream().mapToInt(Integer::intValue).sum() + upperBonus;
            Map<String, Integer> categories = new HashMap<>();
            scores.forEach((category, score) -> categories.put(category.apiKey(), score));
            return new ScoreBoard(categories, upperSubtotal, upperBonus, total);
        }
    }
}

File: backend/src/game/yacht/bot/__tests__/yachtBotGameCompletion.test.ts
import { describe, expect, it } from 'vitest'
import {
  InMemoryRoundStateStore,
  type RoundSubmissionResult,
  RoundSynchronizationService,
  seededDieRoller,
} from '../../../round/index.js'
import {
  SCORE_CATEGORIES,
  ScoreConfirmationService,
  ScoreRoundSubmissionService,
} from '../../../score/index.js'
import { FakeRoundTimer, RecordingBroadcaster } from '../../__tests__/testDoubles.js'
import { YachtTurnActionService } from '../../yachtTurnActionService.js'
import { ExpectimaxYachtBotPolicy } from '../expectimaxYachtBotPolicy.js'
import { LocalYachtBotStrategy } from '../localYachtBotStrategy.js'
import { ScorecardValueEvaluator } from '../scorecardValueEvaluator.js'
import { YachtBotTurnCoordinator } from '../yachtBotTurnCoordinator.js'
import { AccumulatingScoreBoardStore, roomWith } from './botTestDoubles.js'

/**
 * `YachtBotGameCompletionTest` 이식 — **2봇이 12라운드를 사람과 같은 경로로 완주**한다.
 *
 * 이 스위트에서 유일하게 대역이 아닌 것들: 라운드 동기화 서비스, 진짜
 * `YachtTurnActionService`, 진짜 점수 확정 서비스, 진짜 정책·폴백. 대역은 타이머
 * (마감 발화는 2.5가 덮었다)·브로드캐스터·방 스냅샷·점수 저장소뿐이다.
 *
 * 오케스트레이터는 여기서 쓰지 않는다 — 지연 4종을 실시간으로 기다리면 12라운드가
 * 몇 분이 된다. 대신 Java 테스트와 같이 코디네이터를 루프에서 직접 돌린다.
 */

const ROOM = 'room-a'
const GAME = 'game-a'
const BOTS = ['bot-easy', 'bot-hard'] as const

describe('야추 봇 완주', () => {
  it('2봇이 공유 행동 경로로 12라운드를 끝낸다', async () => {
    const rounds = new RoundSynchronizationService(new InMemoryRoundStateStore(), {
      // 서버 RNG를 시드로 고정한다(2.5의 시임) — 실패가 재현 가능해야 한다.
      dieRoller: seededDieRoller(20260814),
    })
    const scoreStore = new AccumulatingScoreBoardStore()
    const scores = new ScoreConfirmationService(scoreStore)
    const rooms = {
      getSnapshot: async () =>
        roomWith(
          GAME,
          BOTS.map((playerId) => ({ playerId, kind: 'BOT' as const })),
        ),
    }
    const submissions = new ScoreRoundSubmissionService<RoundSubmissionResult>(
      rounds,
      scores,
      rooms,
    )
    const timers = new FakeRoundTimer()
    const broadcaster = new RecordingBroadcaster()
    const actions = new YachtTurnActionService({ rounds, timers, broadcaster, submissions })
    const coordinator = new YachtBotTurnCoordinator({
      rounds,
      actions,
      policy: new ExpectimaxYachtBotPolicy(new ScorecardValueEvaluator()),
      strategy: new LocalYachtBotStrategy(),
      rooms,
      scores,
    })

    await rounds.initialize(ROOM, 1, [...BOTS])

    let steps = 0
    for (;;) {
      const state = await rounds.findByRoomId(ROOM)
      expect(state).toBeDefined()
      if (state === undefined || state.finished || steps >= 200) break
      // **모든 스텝이 실제 행동이어야 한다.** 하나라도 무시되면 봇이 멈춘 것이고,
      // 그러면 아래 루프가 200번 돌아 실패한다.
      expect(await coordinator.playIfCurrent({ roomId: ROOM, state })).toBe(true)
      steps += 1
    }

    const finished = await rounds.findByRoomId(ROOM)
    expect(finished?.finished).toBe(true)
    expect(finished?.roundNumber).toBe(12)
    // 라운드당 봇 하나가 최소 2스텝(굴림 + 제출), 최대 6스텝(굴림3 + 킵2 + 제출).
    expect(steps).toBeGreaterThanOrEqual(48)
    expect(steps).toBeLessThanOrEqual(144)

    for (const botId of BOTS) {
      const board = await scoreStore.findScoreBoard(GAME, botId)
      for (const category of SCORE_CATEGORIES) {
        expect(board.categories[category], `${botId}/${category}`).not.toBeNull()
      }
    }
  }, 60_000)
})
